plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs

```diff
@@ -368,6 +364,118 @@ split (Value mp) = (negate (Value neg), Value pos) where
+    -- non-empty lists only consists of empty values.
+    go [] kvs2 = if all (is0 . snd) kvs2 then MatchSuccess else MatchFailure
+    -- Symmetric to the previous case.
+    go kvs1 [] = if all (is0 . snd) kvs1 then MatchSuccess else MatchFailure
+    -- Both spines are non-empty.
+    go kvs1@((k1, v1) : kvs1') kvs2@((k2, v2) : kvs2')
+        -- If keys are equal
+        | k1 == k2 =
+            -- then continue checking equality of spines.
+            case go kvs1' kvs2' of
+                -- If spines are equal, then we can proceed to checking equality of values as by
+                -- this point we've ensured that the keys in the lists match exactly.
+                MatchSuccess -> if structEqV v1 v2 then MatchSuccess else MatchFailure
+                -- If there was a key mismatch down the line, then we cons the values associated
+                -- with @k1@ and @k2@ onto @vvs@ to check their equality after we ensure that keys
+                -- from @kvs1''@ are a permutation of those from @kvs2''@.
+                MatchPartial vvs kvs1'' kvs2'' -> MatchPartial ((v1, v2) : vvs) kvs1'' kvs2''
+                -- A failure stays a failure.
+                MatchFailure -> MatchFailure
+        -- If the keys aren't equal, then maybe the first value is empty, in which case we throw it
+        -- out and proceed.
+        | is0 v1 = go kvs1' kvs2
+        -- Or if the second one is empty, then we throw that one and proceed.
+        | is0 v2 = go kvs1 kvs2'
+        -- Otherwise the keys in the lists have diverged and we return the remaining parts.
+        | otherwise = MatchPartial [] kvs1 kvs2
+
+{-# INLINEABLE eq #-}
+-- | Check equality of two 'Value's. Does not assume orderness of lists within 'Value' or lack of
+-- empty values (such as a token whose quantity is zero or a currencies that has a bunch of such
+-- tokens or no tokens at all), but does assume that no currencies or tokens within a single
+-- currency have multiple entries.
+eq :: Value -> Value -> Bool
+eq (Value (Map.toList -> currs1)) (Value (Map.toList -> currs2)) =
+    -- Check structural equality of the lists first.
+    case matchKVs (Map.all (== 0)) (eqMapVia SortedMap.UnsafeSortedMap) currs1 currs2 of
+        MatchSuccess -> True
+        MatchFailure -> False
+        -- If the lists aren't structurally equal, then convert the remaining parts to 'SortedMap's
+        -- (by sorting first) and check structural equality of the now sorted lists.
+        MatchPartial valPairs currs1' currs2' ->
+            if SortedMap.eqWith
+                    (Map.all (== 0))
+                    (eqMapVia SortedMap.unsafeFromListUnique)
+                    (SortedMap.unsafeFromListUnique currs1')
+                    (SortedMap.unsafeFromListUnique currs2')
+                then
+                    -- Check equality of values that come from the common key prefix of the original
+                    -- lists.
+                    all (uncurry (==)) valPairs
```

Comment:
Here valpairs are maps themselves. Don't you have to do proper sorted equality check (i.e. eqMapVia)?

Reply:
Thank you for spotting the bug, it's fixed now.